src/tstool/tstool.c

```diff
@@ -0,0 +1,620 @@
+{
+	int rc;
+
+	/* Start netlink service */
+	netlink = sfptpd_netlink_init();
+	if (netlink == NULL) {
+		CRITICAL("could not start netlink\n");
+		return EXIT_FAILURE;
+	}
+
+	rc = sfptpd_netlink_scan(netlink);
+	if (rc != 0) {
+		CRITICAL("scanning with netlink, %s\n",
+			 strerror(rc));
+		return rc;
+	}
+
+	/* Wait 5 seconds for initial link table */
+	initial_link_table = sfptpd_netlink_table_wait(netlink, 1, 5000);
+	if (initial_link_table == NULL) {
+		CRITICAL("could not get initial link table, %s\n",
+			 strerror(errno));
+		return EXIT_FAILURE;
+	}
+
+	/* Start clock management */
+	rc = sfptpd_clock_initialise(config, NULL);
+	if (rc != 0)
+		return EXIT_FAILURE;
+
+	/* Start interface management */
+	rc = sfptpd_interface_initialise(config, NULL,
+					 initial_link_table);
+	if (rc != 0)
+		return EXIT_FAILURE;
+
+	return 0;
+}
+
+static void do_finit(void)
+{
+	sfptpd_clock_shutdown();
+	sfptpd_interface_shutdown(config);
+	if (netlink != NULL)
+		sfptpd_netlink_finish(netlink);
+}
+
+static void usage(FILE *stream)
+{
+	fprintf(stream,
+		"syntax: %s [OPTIONS] SUBSYSTEM COMMAND..\n"
+		"\n"
+		"  OPTIONS\n"
+		"        --persistent            Use sfptpd persistent frequency adjustment\n"
+		"        --initial               Perform sfptpd initial clock correction\n"
+		"    -h, --help                  Show usage\n"
+		"    -v, --verbose               Be verbose\n\n"
+		"  CLOCK SUBSYSTEM\n"
+		"    clock list                  List clocks\n"
+		"    clock info CLOCK            Show clock information\n"
+		"    clock get CLOCK             Read clock\n"
+		"    clock step CLOCK OFFSET     Step clock\n"
+		"    clock slew CLOCK PPB        Adjust clock frequency\n"
+		"    clock set_to CLOCK1 CLOCK2  CLOCK1 := CLOCK2\n"
+		"    clock diff CLOCK1 CLOCK2    CLOCK1 - CLOCK2\n"
+		"    clock dedup                 Deduplicate shared phc devices\n\n"
+		"      CLOCK := <phcN> | <ethN> | system\n\n"
+		"  INTERFACE SUBSYSTEM\n"
+		"    interface list              List physical interfaces\n"
+		"    interface info INTF         Show interface information\n"
+		"    interface set_ts TX RX      Set timestamp modes\n\n"
+		"      INTF := <ethN>\n"
+                "      See 'info' response for available TX and RX modes\n",
+		program_invocation_short_name);
+}
+
+static int clock_command(int argc, char *argv[])
+{
+#define MAX_CLOCKS 2
+	struct sfptpd_timespec times[MAX_CLOCKS];
+	struct sfptpd_clock *clocks[MAX_CLOCKS];
+	struct sfptpd_interface *interface;
+	struct sfptpd_clock **all_clocks;
+	const struct clock_command *cmd;
+	const char *command;
+	size_t num_clocks;
+	sfptpd_time_t freq_adj;
+	sfptpd_time_t tick_len;
+	sfptpd_time_t f;
+	int tokens;
+	int rc = 0;
+	int i;
+
+	if (argc < 1) {
+		usage(stderr);
+		return EXIT_FAILURE;
+	}
+
+	command = argv[0];
+
+	for (cmd = clock_cmds;
+	     cmd->tag != CLOCK_CMD_INVALID &&
+	     strcmp(command, cmd->name);
+	     cmd++);
+
+	assert(cmd->clock_args <= MAX_CLOCKS);
```

Comment:
Odd to assert one way, and print error message and usage the other.  Also, wouldn't the assert be compiled out of a customer build?

Reply:
Yup, as Matthew ways, this is just a bonus internal consistency check, not error checking.